src/nccl_ofi_rdma.cpp

```
@@ -3928,6 +3944,67 @@ static inline int recv_comm_insert_send_close_req(nccl_net_ofi_rdma_recv_comm_t
 	return 0;
 }
 
+
+/**
+ * Make progress on a closing recv communicator
+ *
+ * @param r_comm: the communicator to progress
+ * @return: 1 if the recv comm is ready to be destroyed
+ *          0 if the recv comm is not ready to be destroyed
+ *          negative errno code on error
+ */
+static inline int progress_closing_recv_comm(nccl_net_ofi_rdma_recv_comm_t *r_comm)
+{
+	if (r_comm->send_close_req == NULL) {
+		/* Waiting for all ctrls to complete */
+		nccl_net_ofi_mutex_lock(&r_comm->ctrl_counter_lock);
+		uint64_t n_ctrl_sent = r_comm->n_ctrl_sent;
+		bool all_ctrl_msgs_delivered =
+			(r_comm->n_ctrl_delivered == n_ctrl_sent);
+		nccl_net_ofi_mutex_unlock(&r_comm->ctrl_counter_lock);
+
+		if (all_ctrl_msgs_delivered) {
+			/* Send close message, only if we sent any control
+			   messages. Otherwise, destroy the recv comm
+			   immediately */
+			/* TODO: this workaround will not be needed with the new
+			   CM code when data_progress_auto is true */
+			if (n_ctrl_sent == 0) {
+				return 1;
+			}
+
+			int ret = recv_comm_insert_send_close_req(r_comm);
+			if (ret != 0) {
+				return ret;
+			}
+
+			ret = receive_progress(r_comm->send_close_req, true);
+			if (ret != 0) {
+				return ret;
```

Do you think it'd be appropriate to add some WARN logs for these error cases? I know you had moved these calls from `recv_comm_process_all_finalizing` without changing them, I feel it would be beneficial to have more specific warning logs here rather than silently propagate the error up to `recv_comm_process_all_finalizing`, which would then silently propagate the error up to `comm_close_handler`.

---

> Do you think it'd be appropriate to add some WARN logs for these error cases?

We should document somewhere, but the convention has been to only log an error once instead of logging errors all the way up the stack. For example, here I'm assuming that `receive_progress` (which is an internal plugin function) already logged an error if it failed, so I don't need to log another one. Otherwise we end up with a bunch of extra prints of the form:
```
Failed call to "receive_progress"
```
which don't really help the user.

This approach does have its pitfalls, particularly if we forget to put a log in the place the error first occurs.

---

Make sense! Yeah I see that [receive_progress](https://github.com/aws/aws-ofi-nccl/blob/133965713edbb3f094b0145838081b102e0da4f6/src/nccl_ofi_rdma.cpp#L1846) for example doesn't have real warn logs, but the functions it calls (e.g. [post_eager_copy](https://github.com/aws/aws-ofi-nccl/blob/133965713edbb3f094b0145838081b102e0da4f6/src/nccl_ofi_rdma.cpp#L5714)) do have the more granular error logging.